corrigindo fork

sanychiquetti opened the pull request.

sanychiquetti committed:
corrigindo fork

wlsf82 commented on cypress.config.js:
```diff
@@ -0,0 +1,20 @@
+    baseUrl: 'http://localhost:3000',
+    fixturesFolder: false,
+    env: {
+      API_URL: 'http://localhost:3001/',
```
Recomendo ou usar `/` no final de ambas `baseUrl` e API_URL`, ou não usar em ambas.
A consistência é importante para um bom design.

wlsf82 commented on cypress.config.js:
```diff
@@ -0,0 +1,20 @@
+    specPattern: 'frontend/src/**/*.cy.{js,jsx,ts,tsx}',
+    supportFile: 'frontend/cypress/support/component.js'
+  },
+})
```
É uma boa prática deixar uma linha em branco ao final de cada arquivo, para que o código seja melhor renderizado em serviços como o GitHub, onde quando não há tal linha em branco, um ícone é exibido ao final de cada arquivo, informando _No newline at end of file_.

```suggestion
})

```

wlsf82 commented on cypress/.eslintrc.json:
```diff
@@ -0,0 +1,10 @@
+     "cypress/assertion-before-screenshot": "error",
+     "cypress/no-pause": "error"
+   }
+ } 
```
```suggestion
 } 
 
```

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+describe('Listar clientes na aplicação EngageSphere', () => {
```
Que tal?

```suggestion
describe('GET /customers', () => {
```

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+describe('Listar clientes na aplicação EngageSphere', () => {
+  const CUSTOMERS_API_URL = `${Cypress.env('API_URL')}`
```
É uma boa prática (a qual ajuda na legibilidade do código) deixar uma linha em branco entre a definição de variáveis e outras expressões.

```suggestion
  const CUSTOMERS_API_URL = `${Cypress.env('API_URL')}`
  
```

wlsf82 replied:
A propósito, que tal definir o valor abaixo para a variável `CUSTOMERS_API_URL`?

`${Cypress.env('API_URL')}customers`

Dessa forma, a string `customers` não precisa ser repetida em todas as requisições.

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+  const CUSTOMERS_API_URL = `${Cypress.env('API_URL')}`
+  it('Recuperar clientes com sucesso', () => {
+    cy.request('GET', `${CUSTOMERS_API_URL}customers?page=1&limit=10&size=All`).then(
+      (response) => {
```
Veja como seria desestruturando só a propriedade `status` da `response`.

```suggestion
      (({ status })) => {
```

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+  it('Recuperar clientes com sucesso', () => {
+    cy.request('GET', `${CUSTOMERS_API_URL}customers?page=1&limit=10&size=All`).then(
+      (response) => {
+        expect(response.status).to.eq(200)
```
E então, aqui você poderia fazer assim:

```suggestion
        expect(status).to.eq(200)
```

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+      'Very Large Enterprise',
+    ]
+
+    tamanhos.forEach((tamanho) => {
```
👍🏻

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+    })
+  })
+
+  it('Validar filtro por paginação', () => {
```
Este teste está falhando, pois você está tentando usar um `.forEach` em um objeto (`pageInfo `), quando tal método é específico para Arrays.

Aqui vai uma alternativa:

```js
it('Validar filtro por paginação', () => {
  const limiteDeResultadosPorPagina = ['5', '10', '20', '50']
  const totalDePaginasPorLimite = [10, 5, 3, 1] // Supondo que há 50 clientes no banco de dados

  limiteDeResultadosPorPagina.forEach((limite, indice) => {
    cy.request('GET', `${CUSTOMERS_API_URL}customers?page=1&limit=${limite}&size=All`)
      .then(({ body }) => {
        expect(body.customers).to.have.length(limite)
        expect(body.pageInfo.totalPages).to.eq(totalDePaginasPorLimite[indice])
      })
  })
})

```

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+          expect(customer).to.have.property('name')
+          expect(customer).to.have.property('employees')
+          expect(customer).to.have.property('size')
+          expect(customer).to.have.property('address')
```
Faltou a seguinte verificação: `expect(customer).to.have.property('contactInfo')`

wlsf82 commented on cypress/e2e/api/engageShere.cy.js:
```diff
@@ -0,0 +1,144 @@
+      })
+    })
+
+    it('Não fornecer valores obrigatórios nos campos de página, limite e tamanho', () => {
```
Gostei de ter adicionado um cenário que não estava descrito. 👏🏻 

A propósito, que tal mais alguns cenários? Tais como:

- O que acontece ao passar `0` para `page` ou `limit`?
- O que acontece ao filtrar por uma página vazia? (Ex.: `?limit=10&page=6 // supondo que há 50 customers no banco de dados`

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,13 @@
+describe('Testes dos componentes da aplicação Engage Sphere', () => {
+  beforeEach(() => {
+    cy.visit('/')
+  })
```
Para ajudar na legibilidade, recomendo separar o `beforeEach` do `it` por uma linha em branco.

> O mesmo vale para a separação de um `it` do outro.

```suggestion
  })
  
```

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,13 @@
+    cy.visit('/')
+  })
+  it('Mostra uma lista de clientes quando há dados no banco de dados', () => {
+    cy.get('table').should('have.length.greaterThan', 0)
```
Esse teste necessita uma verificação mais adequada.
Essa implementação não garante que dados estejam sendo renderizados na tabela.

Que tal usar o comando `cy.intercept()`, devolvendo ao _frontend_ dados estáticos (conhecidos) por meio de _fixtures_?
Dessa forma, você pode garantir cada dado sendo exibido em cada linha e coluna da tabela.

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,13 @@
+  })
+
+  it('Desativa o campo de entrada texto quando na página de detalhes do cliente', () => {
+    cy.get('table tr td button').first().click()
```
Que tal um seletor mais simples, como:

```suggestion
    cy.get('tbody tr').first().click()
```

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,13 @@
+
+  it('Desativa o campo de entrada texto quando na página de detalhes do cliente', () => {
+    cy.get('table tr td button').first().click()
+    cy.get('[data-testid="name"]').should('have.attr', 'disabled')
```
E aqui, que tal uma verificação mais simples, como:

```suggestion
    cy.get('[data-testid="name"]').should('be.disabled')
```

wlsf82 commented on cypress/e2e/gui/validarCampoTexto.cy.js:
```diff
@@ -0,0 +1,14 @@
+    cy.intercept('GET', '**/customers**').as('getCustomers')
+    cy.visit('/customers?page=1&limit=10&size=All')
+    cy.wait('@getCustomers')
+  })
```
```suggestion
  })
  
```

wlsf82 commented on cypress/e2e/gui/validarCampoTexto.cy.js:
```diff
@@ -0,0 +1,14 @@
+describe('Testes dos componentes da aplicação Engage Sphere', () => {
+  beforeEach(() => {
+    cy.intercept('GET', '**/customers**').as('getCustomers')
+    cy.visit('/customers?page=1&limit=10&size=All')
```
Não sei como esse `cy.visit` está funcionando, visto que você está passando os parâmetros que a API espera.

Aqui deveria ser só o seguinte:

```suggestion
    cy.visit('/')
```

wlsf82 commented on cypress/e2e/gui/validarCampoTexto.cy.js:
```diff
@@ -0,0 +1,14 @@
+    cy.visit('/customers?page=1&limit=10&size=All')
+    cy.wait('@getCustomers')
+  })
+  it('Desativa o campo de entrada de texto quando na página de detalhes do cliente', () => {
```
Este teste está duplicado.
Já um teste para o mesmo cenário no arquivo `cypress/e2e/gui/engageSphere.cy.js`.

sanychiquetti replied:
Eu renomeei o arquivo `cypress/e2e/gui/engageSphere.cy.js` para `cypress\e2e\gui\validarListaClientes.cy.js` para ficar condizente com o teste que ele executa e retirei o teste que já existe no arquivo `cypress/e2e/gui/validarCampoTexto.cy.js`

wlsf82 commented on cypress/e2e/gui/validarFallBackLoading.cy.js:
```diff
@@ -0,0 +1,10 @@
+    cy.visit('/customers?page=1&limit=10&size=All')
+    cy.wait('@getCustomers')
+  })
+  it('Deve exibir a mensagem de carregamento', () => {
```
**Este teste é _flaky_!**

Às vezes ele passa e outras falha.

Para testar estados transitórios, como o componente de _loading..._, é importanto forçar o _delay_ na resposta do servidor, conforme explico no blog post [Como simular a demora em uma requisição com Cypress](https://talkingabouttesting.com/2021/02/26/como-simular-a-demora-em-uma-requisicao-com-cypress/).

Veja uma alternantiva para o mesmo teste, porém, a qual provém resultados determinísticos:

```js
describe('Validar a mensagem de "Loading"', () => {
  beforeEach(() => {
    cy.intercept('GET', '**customers**', { delay: 1000 }).as('getDelayedCustomers')
    cy.visit('/')
  })
  it('Deve exibir a mensagem de carregamento', () => {
    cy.get('#loading').should('be.visible')
    cy.wait('@getDelayedCustomers')
    cy.get('#loading').should('not.exist')
  })
})
```

sanychiquetti replied:
Walmyr, eu não consegui entender uma questão, se o meu teste é para que apareça a mensagem de loading...
Porque o cy.wait fica após a validação de `cy.get('#loading').should('be.visible')`, porque colocar após o assert que realmente quero fazer garante que ele não será flaky?
Para mim fez total sentido para a validação de garantir que o loading... não aparece na tela, pois esperei a requisição acabar e então verifiquei que a mensagem não existe.

wlsf82 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,25 @@
+// ***********************************************
```
Que tal deletar esse arquivo e o `cypress/support/e2e.js`, visto que eles não estão em uso?
Ao fazer isso, lembre-se de atualizar o arquivo de configurações do Cypress (`cypress.config.js`) com a propriedade `supportFile` com valor `false`.

wlsf82 commented on package.json:
```diff
@@ -1,8 +1,12 @@
 {
-  "name": "EngageSphere-Test-Design-Masterclass-TAT",
+  "name": "engagesphere-test-design-masterclass-tat",
```
Algum motivo para mudar o nome do projeto?

```suggestion
  "name": "EngageSphere-Test-Design-Masterclass-TAT",
```

wlsf82 commented on package.json:
```diff
@@ -29,5 +33,20 @@
+  "bugs": {
+    "url": "https://github.com/wlsf82/EngageSphere-Test-Design-Masterclass-TAT/issues"
+  },
+  "main": "index.js",
```
Eu diria que o `main` nesse caso é o `cypress.config.js`.

wlsf82 reviewed (changes requested):
Bom trabalho @sanychiquetti! 👏🏻 

No entanto, deixei alguns comentários e sugestões para novas melhorias.

Além disso, estou ansioso para ver como vai se sair com os testes de componentes.

Precisando de ajuda, conte comigo!

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,13 @@
+describe('Testes dos componentes da aplicação Engage Sphere', () => {
```
Que tal só:

```suggestion
describe('EngageSphere', () => {
```

?

sanychiquetti committed:
Aplicando correções nos testes:
Deixando uma linha a mais após o término do cógigo, renomeando arquivo engageShere para o nome validarListaClientes para deixar mais espefícico, e tb apagando teste duplicado.
Inserindo a / na base_Url e refatorando o cy.visit dos testes de GUI.

sanychiquetti committed:
tornando o teste 'validar a mensagem de Loading' determinístico

sanychiquetti committed:
deletando arquivos da pasta support que não estão em uso

sanychiquetti committed:
setando a propriedade main do arquivo package.json para o valor cypress.config.js

sanychiquetti committed:
corrigindo teste de API de validar filtro por paginação

sanychiquetti committed:
acrescentando assert de validar a propriedade de contactInfo no teste de 'Validar as propeiedades customer e pagInfo da resposta'

sanychiquetti committed:
acrescentando o teste de validar filtro com valor igual a 0 no context de'Validar solicitações inválidas'

sanychiquetti committed:
renomeando o describe do arquivo 'valida CampoTexto' para ficar mais condinzente com o teste em questão
E para melhora a legibilidade do código inserindo uma linha entre as funções

sanychiquetti committed:
corrigindo endereço cy.visit e a busca de elemento arquivo 'validarCampoTexto'

sanychiquetti committed:
interceptando a requisição dos testes 'validatCampoTexto'e 'validarListaClientes'

sanychiquetti committed:
enviando testes corrigidos para validação

sanychiquetti committed:
inserindo os testes de acessibilidade
